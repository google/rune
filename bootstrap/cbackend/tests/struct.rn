struct Point {
  x: Int = 0i32
  y: Int = 0i32
}

p = Point(1i32, 2i32)
println p
println "p.x=", p.x
println "p.y=", p.y
println "p[0]=", p[0]
println "p[1]=", p[1]
p.x = 3i32
println p
p[1] = 3i64
println p

struct TwoPoints {
  s1: Point
  s2: Point
}

tp = TwoPoints(Point(1i32, 2i32), Point(3i32, 4i32))

println "s1.x = ", tp.s1.x
println "s1.y = ", tp.s1.y
println "s2.x = ", tp.s2.x
println "s2.y = ", tp.s2.y

/* Polymorphic structs */
struct Person {
  id
  age
};

a = Person("Bob", 32)
println "Person a = ", a
b = Person(123, "old")
println "Person b = ", b


/* TODO: handle named parameters
q = Point(
    x = 3i64,
    y = 4i64,
)
println q
*/

/* TODO: handle default parameters
r = Point()
println r
*/

