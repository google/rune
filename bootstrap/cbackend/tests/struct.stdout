Point(1, 2)
p.x=1
p.y=2
p[0]=1
p[1]=2
Point(3, 2)
Point(3, 3)
s1.x = 1
s1.y = 2
s2.x = 3
s2.y = 4
Person a = Person_string_u64("Bob", 32)
Person b = Person_u64_string(123, "old")
